catalog/security/catalog-security-xmlattributeplugin/src/main/java/org/codice/ddf/catalog/security/policy/xml/XmlAttributeSecurityPolicyPlugin.java

```diff
@@ -0,0 +1,281 @@
+
+            while (xmlStreamReader.hasNext()) {
+                int event = xmlStreamReader.next();
+                if (event == XMLStreamConstants.START_ELEMENT) {
+                    String localName = xmlStreamReader.getLocalName();
+
+                    if (xmlElements.contains(localName)) {
+                        LOGGER.debug("Parsing security attribute.");
+                        parseSecurityBlock(securityMap, xmlStreamReader);
+                    }
+                }
+            }
+        } catch (XMLStreamException e) {
+            //if this happens and message redacting is enabled, the message will be excluded from results
+            LOGGER.warn("Unable to parse security from XML metadata.", e);
+        } finally {
+            if (xmlStreamReader != null) {
+                try {
+                    xmlStreamReader.close();
+                } catch (XMLStreamException e) {
+                    //ignore
+                }
+            }
+        }
+
+        return securityMap;
+    }
+
+    /**
+     * Parses the security element in the metadata
+     *
+     * @param xmlStreamReader xml stream
+     */
+    private void parseSecurityBlock(Map<String, Set<String>> securityMap,
+            XMLStreamReader xmlStreamReader) {
+        LOGGER.debug("Parsing metacard security block");
+        int numAttrs = xmlStreamReader.getAttributeCount();
+
+        for (int i = 0; i < numAttrs; i++) {
+            String name = xmlStreamReader.getAttributeLocalName(i);
+            if (getSecurityAttributeUnions().contains(name)) {
+                LOGGER.debug("Found {} in metacard", name);
+                if (!securityMap.containsKey(name)) {
+                    securityMap.put(name, new HashSet<>());
+                }
+                buildSecurityAttribute(securityMap.get(name), xmlStreamReader.getAttributeValue(i));
+            } else if (getSecurityAttributeIntersections().contains(name)) {
+                if (!securityMap.containsKey(name)) {
+                    securityMap.put(name, new HashSet<>());
+                }
+                buildIntersectionAttribute(securityMap.get(name),
+                        xmlStreamReader.getAttributeValue(i));
+            }
+        }
+
+    }
+
+    private void buildSecurityAttribute(Set<String> builderSet, String attributeValue) {
+        StringTokenizer tokenizer = new StringTokenizer(attributeValue, " ", false);
+        while (tokenizer.hasMoreElements()) {
+            String value = tokenizer.nextToken();
+            LOGGER.debug("Adding {} to set", value);
+            builderSet.add(value);
+        }
+    }
+
+    private void buildIntersectionAttribute(Set<String> builderSet, String attributeValue) {
+        StringTokenizer tokenizer = new StringTokenizer(attributeValue, " ", false);
+        Set<String> attributeSet = new HashSet<>();
+
+        boolean empty = builderSet.isEmpty();
+
+        // Create the tmpSet
+        while (tokenizer.hasMoreElements()) {
```

Comment:
I think this would be more efficient like this:
```
if (builderSet.isEmpty()) {
    buildSecurityAttribute(builderSet, attributeValue);
} else {
    // do the tmpSet/builderSet work as is
```